src/jit/gentree.cpp

```diff
@@ -6814,105 +7077,113 @@ void Compiler::gtBlockOpInit(
 
-        GenTreePtr srcChild = srcOrFillVal->gtGetOp1();
-        GenTreePtr dstChild = dst->gtGetOp1();
-
-        if (dstChild->OperIsLocal() && varTypeIsStruct(dstChild) && srcChild->OperGet() == GT_SIMD &&
-            varTypeIsSIMD(srcChild))
+        GenTree* src = srcOrFillVal;
+        if (src->OperIsIndir() && (src->AsIndir()->Addr()->OperGet() == GT_ADDR))
         {
-            unsigned   lclNum                = dst->gtGetOp1()->AsLclVarCommon()->GetLclNum();
-            LclVarDsc* lclVarDsc             = &lvaTable[lclNum];
-            lclVarDsc->lvUsedInSIMDIntrinsic = true;
+            src = src->AsIndir()->Addr()->gtGetOp1();
+        }
+        if (src->OperGet() == GT_SIMD)
+        {
+            if (dst->OperIsBlk() && (dst->AsIndir()->Addr()->OperGet() == GT_ADDR))
+            {
+                dst = dst->AsIndir()->Addr()->gtGetOp1();
+            }
+
+            if (dst->OperIsLocal() && varTypeIsStruct(dst))
+            {
+                unsigned   lclNum                = dst->AsLclVarCommon()->GetLclNum();
+                LclVarDsc* lclVarDsc             = &lvaTable[lclNum];
+                lclVarDsc->lvUsedInSIMDIntrinsic = true;
+            }
         }
     }
 #endif // FEATURE_SIMD
 }
 
 //------------------------------------------------------------------------
-// gtNewBlkOpNode: Creates an InitBlk or CpBlk node.
+// gtNewBlkOpNode: Creates a GenTree for a block (struct) assignment.
 //
 // Arguments:
-//    oper          - GT_COPYBLK, GT_INITBLK or GT_COPYOBJ
 //    dst           - Destination or target to copy to / initialize the buffer.
-//    srcOrFillVall - Either the source to copy from or the byte value to fill the buffer.
-//    sizeOrClsTok  - The size of the buffer or a class token (in the case of CpObj).
+//    srcOrFillVall - the size of the buffer to copy/initialize or zero, in the case of CpObj.
+//    size          - The size of the buffer or a class token (in the case of CpObj).
 //    isVolatile    - Whether this is a volatile memory operation or not.
+//    isCopyBlock   - True if this is a block copy (rather than a block init).
 //
 // Return Value:
 //    Returns the newly constructed and initialized block operation.
-
-GenTreeBlkOp* Compiler::gtNewBlkOpNode(
-    genTreeOps oper, GenTreePtr dst, GenTreePtr srcOrFillVal, GenTreePtr sizeOrClsTok, bool isVolatile)
+//
+// Notes:
+//    If size is zero, the dst must be a GT_OBJ with the class handle.
+//
+GenTree* Compiler::gtNewBlkOpNode(
+    GenTreePtr dst, GenTreePtr srcOrFillVal, unsigned size, bool isVolatile, bool isCopyBlock)
 {
-    GenTreeBlkOp* result = new (this, oper) GenTreeBlkOp(oper);
-    gtBlockOpInit(result, oper, dst, srcOrFillVal, sizeOrClsTok, isVolatile);
+    if (isCopyBlock)
+    {
+        srcOrFillVal->gtFlags |= GTF_DONT_CSE;
+        if (srcOrFillVal->OperIsIndir() && (srcOrFillVal->gtGetOp1()->gtOper == GT_ADDR))
+        // && (srcOrFillVal->gtGetOp1()->gtGetOp1()->TypeGet() == dst->TypeGet()))
+        {
```

Comment:
Because it turns out that the existing code just retypes the source in that case - but I failed to eliminate this after successfully matching the former behavior.

---
In reply to: [77258955](https://github.com/dotnet/coreclr/pull/7027#discussion_r77258955) [](ancestors = 77258955)